File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/historicosalarial/OcorrenciaDoHistoricoSalarialOptimizerListSearch.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.negocio.dominio.historicosalarial;

import br.jus.trt8.pjecalc.base.comum.Competencia;
import br.jus.trt8.pjecalc.negocio.comum.OcorrenciaIterator;
import br.jus.trt8.pjecalc.negocio.comum.OptimizerListSearch;
import br.jus.trt8.pjecalc.negocio.dominio.historicosalarial.OcorrenciaDoHistoricoSalarial;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class OcorrenciaDoHistoricoSalarialOptimizerListSearch
extends OptimizerListSearch<Competencia, OcorrenciaDoHistoricoSalarial> {
    private Map<Competencia, OcorrenciaIterator<OcorrenciaDoHistoricoSalarial>> map;

    @Override
    public OptimizerListSearch<Competencia, OcorrenciaDoHistoricoSalarial> init(Collection<OcorrenciaDoHistoricoSalarial> collection) {
        this.map = new HashMap<Competencia, OcorrenciaIterator<OcorrenciaDoHistoricoSalarial>>();
        for (OcorrenciaDoHistoricoSalarial ocorrencia : collection) {
            Competencia competencia = Competencia.getInstance(ocorrencia.getDataOcorrencia());
            OcorrenciaIterator<OcorrenciaDoHistoricoSalarial> iterator = this.map.get(competencia);
            if (iterator == null) {
                iterator = new OcorrenciaIterator<OcorrenciaDoHistoricoSalarial>(ocorrencia);
                this.map.put(competencia, iterator);
                continue;
            }
            iterator.add(ocorrencia);
        }
        return this;
    }

    @Override
    public Iterator<OcorrenciaDoHistoricoSalarial> search(Competencia key) {
        OcorrenciaIterator<OcorrenciaDoHistoricoSalarial> iterator = this.map.get(key);
        if (iterator != null) {
            iterator.gotoFirst();
        }
        return iterator;
    }
}

File: src/lib/pjecalc/core/dominio/historicosalarial/ocorrencia-do-historico-salarial-optimizer-list-search.ts
/**
 * Porte 1:1 de OcorrenciaDoHistoricoSalarialOptimizerListSearch.java (44 linhas).
 *
 * Otimizador de busca — agrupa ocorrências do histórico salarial por competência
 * (mês/ano) para acesso O(1) via Map.
 *
 * Ref: pjecalc-fonte/.../dominio/historicosalarial/OcorrenciaDoHistoricoSalarialOptimizerListSearch.java
 */
import type { OcorrenciaDoHistoricoSalarial } from './ocorrencia-do-historico-salarial';

/** Chave de competência: ano-mês. */
function toCompetenciaKey(date: Date | null): number {
  if (!date) return -1;
  return date.getFullYear() * 100 + date.getMonth();
}

export class OcorrenciaDoHistoricoSalarialOptimizerListSearch {
  private map: Map<number, OcorrenciaDoHistoricoSalarial[]> = new Map();

  init(collection: Iterable<OcorrenciaDoHistoricoSalarial>): this {
    this.map = new Map();
    for (const oc of collection) {
      const key = toCompetenciaKey(oc.getDataOcorrencia());
      if (key === -1) continue;
      const existing = this.map.get(key);
      if (existing) existing.push(oc);
      else this.map.set(key, [oc]);
    }
    return this;
  }

  /** Busca pela competência (mês/ano de uma data). */
  search(competenciaDate: Date | null): Iterator<OcorrenciaDoHistoricoSalarial> | null {
    if (!competenciaDate) return null;
    const key = toCompetenciaKey(competenciaDate);
    const list = this.map.get(key);
    if (!list) return null;
    let idx = 0;
    return {
      next(): IteratorResult<OcorrenciaDoHistoricoSalarial> {
        if (idx < list.length) return { value: list[idx++], done: false };
        return { value: undefined, done: true };
      },
    };
  }

  /** Como o Java: retorna lista direta para manipulação. */
  searchList(competenciaDate: Date | null): OcorrenciaDoHistoricoSalarial[] {
    if (!competenciaDate) return [];
    return this.map.get(toCompetenciaKey(competenciaDate)) ?? [];
  }
}
